File: leetcode/sorting/mergesort.cpp
#include<iostream>
#include<vector>
#include<math.h>

using namespace std;

void mergesort(vector<int>* arr);

int main()
{
	vector<int> arry{6,5,3,1,8,7,2,4};
  mergesort(&arry); 
  cout << "Result: " << endl;
  for(vector<int>::iterator it = arry.begin(); it != arry.end(); ++it)
      cout << *it << ", ";
    cout << endl;
}


void mergesort(vector<int>* arr)
{
  if(arr->size() > 1)
  {
    unsigned mid = ceil(arr->size() / 2);
    vector<int> leftHalf;
    vector<int> rightHalf;
    for(unsigned i = 0; i < mid; ++i)
      leftHalf.push_back(arr->at(i));
    for(unsigned i = mid; i < arr->size(); ++i)
      rightHalf.push_back(arr->at(i));
    cout << "\nsplitting " << endl;
    cout << "leftHalf: " << endl;
    for(vector<int>::iterator it = leftHalf.begin(); it != leftHalf.end(); ++it)
      cout << *it << ", ";
    
    cout << "\nrightHalf: " << endl;
    for(vector<int>::iterator it = rightHalf.begin(); it != rightHalf.end(); ++it)
      cout << *it << ", ";
    cout << endl;
    mergesort(&leftHalf);
    mergesort(&rightHalf);
    
    unsigned i = 0;
    unsigned j = 0;
    unsigned k = 0;
    
    while(i < leftHalf.size() && j < rightHalf.size())
    {
      if(leftHalf[i] < rightHalf[j])
      {
        arr->at(k) = leftHalf[i];
        ++i;
      }
      else
      {
        arr->at(k) = rightHalf[j];
        ++j;
      }
      ++k;
    }
    
    while(i < leftHalf.size())
    {
      arr->at(k) = leftHalf[i];
      ++i;
      ++k;
    }

    while(j < rightHalf.size())
    {
      arr->at(k) = rightHalf[j];
      ++j;
      ++k;
    }

    cout << " merging " << endl;
    for(vector<int>::iterator it = arr->begin(); it != arr->end(); ++it)
      cout << *it << ", ";
    cout << endl;
    
  }
}

File: leetcode/sorting/mergesort.py
import math

def mergesort(arr, n):
  if n > 1:
    leftarr = arr[ : int(math.ceil(n / 2))]
    rightarr = arr[int(math.ceil(n / 2)) : ]
    
    mergesort(leftarr, len(leftarr))
    mergesort(rightarr, len(rightarr))
    
    i = 0
    j = 0
    k = 0
    print 'left half ', leftarr
    print 'right half ', rightarr
    print arr
    while i < len(leftarr) and j < len(rightarr):
      if leftarr[i] < rightarr[j]:
	arr[k] = leftarr[i]
	i += 1
      else:
	arr[k] = rightarr[j]
	j += 1
      k += 1
    print 'after merging', arr
    print i, ' ', j
    # either i or j will reach its maximum value at this stage
    # so the order of the following while loops does not matter
    
    # following while loops are used to fill in the original array (arr)
    # with unmoved (sorted) numbers
    while i < len(leftarr):
      arr[k] = leftarr[i]
      i += 1
      k += 1
    
    while j < len(rightarr):
      arr[k] = rightarr[j]
      j += 1
      k += 1
    print 'current arr', arr
      
      
      
      
    """  
    if arr1[0] > arr2[0]:
      return mergesort(arr2, len(arr2)) + mergesort(arr1, len(arr1))
    else:
      return mergesort(arr1, len(arr1)) + mergesort(arr2, len(arr2))
    """
def main():
  arr = [6,5,3,1,8,7,2,4]
  n = len(arr)
  mergesort(arr, n)
  print arr
  
main()
  